downloader/artifact_downloader.go

```diff
@@ -97,56 +124,72 @@ func (ad *ConcurrentArtifactDownloader) downloadFile(targetDir, fileName, downlo
 	return fileFullPath, nil
 }
 
-func (ad *ConcurrentArtifactDownloader) downloadDirectory(targetDir, fileName, downloadURL string) (string, error) {
+func (ad *ConcurrentArtifactDownloader) downloadAndExtractZipArchive(targetDir, fileName, downloadURL string) (string, error) {
+	tmpDir, err := pathutil.NormalizedOSTempDirPath("pull-intermediate-files")
+	if err != nil {
+		return "", err
+	}
+
+	fileFullPath, err := ad.downloadFile(tmpDir, fileName, downloadURL)
+	if err != nil {
+		return "", err
+	}
+
+	dirName := strings.TrimSuffix(fileName, filepath.Ext(fileName))
+	dirPath := filepath.Join(targetDir, dirName)
+
+	if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
+		return "", err
+	}
+
+	if err := ad.extractZipArchive(fileFullPath, dirPath); err != nil {
+		return "", err
+	}
+
+	return dirPath, nil
+}
+
+func (ad *ConcurrentArtifactDownloader) downloadAndExtractTarArchive(targetDir, fileName, downloadURL string) (string, error) {
 	client := retry.NewHTTPClient()
 
 	resp, err := client.Get(downloadURL)
 	if err != nil {
 		return "", err
 	}
 
-	fmt.Println("targetDir: ", targetDir)
-
 	dirName := strings.TrimSuffix(fileName, filepath.Ext(fileName))
 	dirPath := filepath.Join(targetDir, dirName)
 
 	if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
 		return "", err
 	}
 
-	if err := ad.extractArchive(resp.Body, dirPath); err != nil {
+	if err := ad.extractTarArchive(resp.Body, dirPath); err != nil {
 		return "", err
 	}
 
 	if err := resp.Body.Close(); err != nil {
-		log.Warnf("Failed to close response body: %s", err)
+		ad.Logger.Warnf("Failed to close response body: %s", err)
 	}
 
 	return dirPath, nil
 }
 
-func (ad *ConcurrentArtifactDownloader) download(jobs <-chan downloadJob, results chan<- ArtifactDownloadResult) {
-	for j := range jobs {
-		var fileFullPath string
-		var err error
+func (ad *ConcurrentArtifactDownloader) extractZipArchive(archivePath string, targetDir string) error {
```

Comment:
The `extractZipArchive` and `extractTarArchive` share a big code piece which we could move out:
```
func (ad *ConcurrentArtifactDownloader) extractZipArchive(archivePath string, targetDir string) error {
	cmd := ad.CommandFactory.Create("unzip", []string{archivePath}, &command.Opts{Dir: targetDir})
	return ad.runExtractionCommand(cmd)
}

func (ad *ConcurrentArtifactDownloader) extractTarArchive(r io.Reader, targetDir string) error {
	tarArgs := []string{
		"-x",      // -x: extract files from an archive: https://www.gnu.org/software/tar/manual/html_node/extract.html#SEC25
		"-f", "-", // -f "-": reads the archive from standard input: https://www.gnu.org/software/tar/manual/html_node/Device.html#SEC155
	}
	cmd := ad.CommandFactory.Create("tar", tarArgs, &command.Opts{
		Stdin: r,
		Dir:   targetDir,
	})

	return ad.runExtractionCommand(cmd)
}

func (ad *ConcurrentArtifactDownloader) runExtractionCommand(cmd command.Command) error {
	if out, err := cmd.RunAndReturnTrimmedCombinedOutput(); err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			return fmt.Errorf("command failed with exit status %d (%s): %w", exitErr.ExitCode(), cmd.PrintableCommandArgs(), errors.New(out))
		}
		return fmt.Errorf("%s failed: %w", cmd.PrintableCommandArgs(), err)
	}

	return nil
}
```

Reply:
updated in https://github.com/bitrise-steplib/bitrise-step-pull-intermediate-files/pull/3/commits/07c082ff85e62bf5be80f8698e9e5595347a0122